ldp/graph/common_ops.py

```diff
@@ -283,6 +297,35 @@ async def forward(
 
         return result.messages[0]
 
+    async def _call_single_and_maybe_validate(
+        self, model: LLMModel, num_retries: int, **kwargs
+    ) -> LLMResult:
+        if not self.response_validator:
+            # If a response validator is not supplied, then we should not do any retries here - leave
+            # that for LiteLLM to handle.
+            return await model.call_single(**kwargs)
+
+        @tenacity.retry(
+            retry=tenacity.retry_if_exception_type(ResponseValidationError),
+            # num_retries+1 because the first call is not a retry
+            stop=tenacity.stop_after_attempt(num_retries + 1),
+            wait=tenacity.wait_fixed(1),
+        )
+        async def call_and_validate() -> LLMResult:
+            result = await model.call_single(**kwargs)
+
+            validated = cast(Callable, self.response_validator)(result)
+            if is_coroutine_callable(self.response_validator):
+                validated = await validated
```

Comment:
So the design pattern for validators is:
- Return `None` if the validator passes
- Blows up if the validator fails

Think about how Pydantic `BaseModel` validations work. If they pass, the code silently proceeds with instantiation, otherwise a `ValidationError` is thrown

Can we somehow modify this code to be:

```python
try:
    self.response_validator(result)
except Exception as exc:
    raise ResponseValidationError() from exc
```

Reply:
Okay we can just rethink this later, it's fine with me for now